pywhy_stats/categorical.py

```diff
@@ -0,0 +1,293 @@
+        probability of observing the computed chi-square statistic (or an even
+        higher value), given the null hypothesis that X \u27C2 Y | Zs is True.
+        If boolean = True, returns True if the p_value of the test is greater
+        than `significance_level` else returns False.
+
+    See Also
+    --------
+    scipy.stats.power_divergence
+
+    References
+    ----------
+    .. footbibliography::
+    """
+    if X.ndim != 1 or Y.ndim != 1:
+        raise ValueError("X and Y should be 1-D arrays.")
+
+    # ensure we use numpy arrays
+    X = np.asarray(X)
+    Y = np.asarray(Y)
+
+    # Check if all elements are integers
+    if not np.issubdtype(X.dtype, np.integer):
+        le = LabelEncoder()
+        X = le.fit_transform(X)
+        # warn("Converting X array to categorical array using scikit-learn's LabelEncoder.")
+
+    # Check if all elements are integers
+    if not np.issubdtype(Y.dtype, np.integer):
+        le = LabelEncoder()
+        Y = le.fit_transform(Y)
+        # warn("Converting Y array to categorical array using scikit-learn's LabelEncoder.")
+
+    for name, arr in zip(["X", "Y"], [X, Y]):
+        # Check if the number of unique values is reasonably small
+        unique_values = np.unique(arr)
+        num_unique_values = len(unique_values)
+        if num_unique_values > 64:  # Adjust the threshold as needed
+            raise RuntimeError(
+                f"There are {num_unique_values} unique categories for {name}. "
+                f"This is likely an error."
+            )
+
+    # Step 1: Do a simple contingency test if there are no conditional variables.
+    if Z is None:
+        # Compute the contingency table
+        observed_xy, _, _ = np.histogram2d(X, Y, bins=(np.unique(X).size, np.unique(Y).size))
+        chi, p_value, dof, expected = stats.chi2_contingency(observed_xy, lambda_=lambda_)
+
+    # Step 2: If there are conditionals variables, iterate over unique states and do
+    #         the contingency test.
+    else:
+        import pandas as pd
+
+        chi = 0
+        dof = 0
+
+        Z = np.asarray(Z)
+        if Z.ndim == 1:
+            Z = Z.reshape(-1, 1)
+
+        # XXX: needed when converting to only numpy API
+        # Check if all elements are integers
+        if not np.issubdtype(Z.dtype, np.integer):
+            le = LabelEncoder()
+            for idx in range(Z.shape[1]):
+                Z[:, idx] = le.fit_transform(Z[:, idx])
+            # warn("Converting Z array to categorical array using scikit-learn's LabelEncoder.")
+
+        # check number of samples relative to degrees of freedom
+        # assuming no zeros
+        s_size = Z.shape[1]
+        n_samples = Z.shape[0]
+
+        levels_x = len(np.unique(X))
+        levels_y = len(np.unique(Y))
+        dof_check = (
+            (levels_x - 1)
+            * (levels_y - 1)
+            * np.prod([len(np.unique(Z[:, i])) for i in range(s_size)])
+        )
+        n_samples_req = 10 * dof_check
```

Comment:
Where does the `10` come from?

Reply:
I think used a similar factor in a f-test :) Not sure if exposing it as a parameter makes sense, since a user could just adjust it such that it runs.
What about, we just give out a warning instead of a runtime error. Strictly speaking, it is not an error and also not a hard assumption. We would just expect less accurate results if there are less samples, i.e., it shouldn't cause a runtime error if a user (on their own risk) calls it with too few samples.